Add Kafka documentation (#154)
restatedev/documentation: docs/services/sdk/kafka.mdx

    // Other rpc or event handlers
});

restate.createServer().bindKeyedRouter("myRouter", router).listen(8080);
```

The Kafka record key must be serialized as valid UTF-8 string, otherwise the event won't be delivered.

You can mix event handlers with other event handlers or rpc handlers. For example, let's assume you're implementing a `profile` router to handle and store user profiles in Restate, where the key is the profile unique identifier.
You can implement an rpc method `get` to retrieve the profile and attach an event handler `updateProfile` to consume events from Kafka to update the profile.

Events will be delivered in order with regard to the event key.

</TabItem>
<TabItem value="grpc" label="gRPC API">

### Handling events within keyed services

You can handle events within keyed services, where the key of the Kafka record matches the key used by the other service input messages:

```protobuf
syntax = "proto3";
import "dev/restate/ext.proto";
import "dev/restate/events.proto";

service MyService {
    option (dev.restate.ext.service_type) = KEYED;

    rpc Handle(dev.restate.StringKeyedEvent) returns (google.protobuf.Empty);
    // Other methods
}

> ```

Author: Can do, i didn't add it to avoid making the page too long and because it is not different to any other code-generated method.

The Kafka record key must be serialized as valid UTF-8 string, otherwise the event won't be delivered.

You can mix methods receiving events with regular rpc methods. For example, let's assume you're implementing a `Profile` service to handle and store user profiles in Restate, where the key is the profile unique identifier.
You can implement a method `Get` to retrieve the profile and attach an event handler `UpdateProfile` to consume events from Kafka to update the profile.

Events will be delivered in order with regard to the event key.

### Handling events with a custom key

> You can handle events where the Kafka record key is available but is not a String, for example when the key is a composed key, serialized using Avro or Protobuf.

Author: > parses the string key into the composite type, and then generates a one-way call to the other service.

The event handler doesn't necessarily need do implement that. Inside that event handler, the user can do whatever they want (like any other rpc method). Some use cases I have in mind are for example to just apply some stateless transformation and send the event back to kafka or some other queue using a side effect, or perhaps to convert a kafka message into an http invocation to a 3rd party service, or fan out the service to multiple restate services after validating it.

To do so, you need to handle events in an ad-hoc service that process these events and uses them to generate one-way call to other services, or side effects.

For example:

```protobuf
syntax = "proto3";
import "dev/restate/ext.proto";
import "dev/restate/events.proto";

service MyService {
    option (dev.restate.ext.service_type) = KEYED;

    rpc Handle(dev.restate.KeyedEvent) returns (google.protobuf.Empty);
}
```

Events will be delivered in order with regard to the event key.

### Handling events without key

> You can handle events where the Kafka record key is either not available, or not relevant, and you need a transformation of the event payload before consuming the event.

Author: See above comment.

Similarly to the aforementioned case, you need an ad-hoc service to process those events:

```protobuf
syntax = "proto3";
import "dev/restate/ext.proto";
import "dev/restate/events.proto";

service MyService {
    option (dev.restate.ext.service_type) = KEYED;

    rpc Handle(dev.restate.Event) returns (google.protobuf.Empty);
}
```

> Events will be delivered in order with regard to the event key. If you don't need ordering guarantees and you want to process events as fast as possible, then use the service type `UNKEYED`.

Author: sorry, bad copy paste.

    }

    // Other service methods
}
```

The Kafka record key must be a valid UTF-8 string, otherwise Restate won't be able to deliver the event to the service.

You can mix methods receiving events with regular RPC methods. For example, assume you're implementing a `Profile` service to handle and store user profiles in Restate, where the key is the profile's unique identifier.
In the same service, you can implement a  `Get` method to retrieve the profile, and you can attach an `UpdateProfile` event handler to consume events from Kafka to update the profile.

For each event key (= Kafka record key), the events are delivered in the order in which they arrived on the topic partition.

### Handling generic events from Kafka in order

You can handle events where the Kafka record key is available but is not necessarily a string. To do so, you need to handle events in an ad-hoc service that process these events:

For example:

```protobuf
syntax = "proto3";
import "dev/restate/ext.proto";
import "dev/restate/events.proto";

service MyService {
    option (dev.restate.ext.service_type) = KEYED;

    rpc Handle(dev.restate.KeyedEvent) returns (google.protobuf.Empty);
}
```

> For each event key (= Kafka record key), thanks to Restate [`KEYED` service type](/services/service_type), the events are delivered in the order in which they arrived on the topic partition.

Reviewer: For keyed services, events with the same key are delivered in the order in which they arrived on the topic partition.

Reviewer: ```suggestion
For keyed services, events with the same key are delivered in the order in which they arrived on the topic partition.
```

> These events handler can be useful when you need to pre-process an event before sending it to one or more Restate services, such as when the key is serialized using a custom format, or when the destination of the event is encoded in the payload.

Reviewer: ```suggestion
Event handlers can be useful when you need to pre-process an event before sending it to one or more Restate services, such as when the key is serialized using a custom format, or when the destination of the event is encoded in the payload.
```
